(CFACT-257) Add load_averages facts

mattthias opened the pull request:
This commit adds a fact `load_averages` which contains the number of
processes in the system run queue averaged over time ( over 1 minute, 5 minutes and 15 minutes).

The three load values for 1, 5 and 15 minutes are also available as
facts `load_average_1m`, `load_average_5m` and `load_average_15m`.

kylog commented:
@mattthias thanks for the contribution! This looks great.

Two high-level comments before diving into code review proper:

1) We're trying to move away from built-in facts being in two places, i.e. *both* as top-level 'flat' facts *and* inside a structured fact, in favor of just having structured facts. For the legacy 'flat' facts this will require some discussion and/or a deprecation period. But for brand new facts like this one, we'd rather have a single source of truth for a particular fact and that be in a structured fact.

2) You've made this a linux-only fact, but I *think* it would be reasonable to make it a 'posix' fact, so that it is available to all *nix platforms (osx, solaris, *bsd, aix when we get to aix). Strictly speaking, I don't think `getloadavg` is defined by POSIX, but as a practical matter it seems like all the usual platforms implement it (I say this strictly based on some quick web searches, not direct experience).

Thoughts on the above two considerations?

mattthias commented:
Hi @kylog,

according 1) i'am fine with having only the structured fact. I just looked at some of the other facts :). 
So i should remove the load_average_* facts? 

2) Having this as posix fact was my first idea but i also read that `getloadavg` isn't posix so i moved it to the linux section. 

So, i'll rework this to things: 
 - remove the flat facts load_average_..
 - move to the fact posix section  

Best wishes,
Matthias

mattthias commented:
Hi @kylog ,

i removed the flat facts and moved the fact to the posix section.  Please review :-). 
When everything is in a "mergeable" state i would create a new pull-request with only one commit in it the keep the history more clean. Ok?

best wishes,
Matthias

peterhuene commented:
Hi @mattthias, feel free to rebase or amend your commits and force push to your repository's branch to update this PR so that you don't have to create a new one.  That's generally what we do when we want to squash commits.  Thanks again for the contribution; we'll review your changes soon!

peterhuene commented:
@mattthias Looks like you accidentally edited a commit that has already been merged to master when you squashed the commits.  Could you do a "git pull --rebase upstream master" (assuming `https://github.com/puppetlabs/cfacter.git` has been added as the "upstream" remote) to ensure that only your commit is in the PR?  Thanks!

mattthias commented:
@peterhuene,  ahh, sorry! I fixed that... :-)

peterhuene commented on lib/inc/internal/facts/posix/load_average_resolver.hpp:
```diff
@@ -0,0 +1,23 @@
+        /**
+         * Gets the load averages (for 1, 5 and 15 minutes period).
+         */
+        virtual int get_load_averages(double loads[]) override;
```
Rather than taking a C-style array, I think this member should return a `std::tuple<double, double, double>` and take no parameters.  Also, missing a `@return` doc comment.

peterhuene replied:
Although in the case where load averages could not be obtained, perhaps this should return a `boost::optional<std::tuple<double, double, double>>`.

peterhuene commented on lib/inc/internal/facts/resolvers/load_average_resolver.hpp:
```diff
@@ -0,0 +1,35 @@
+         * Get the system load averages (1, 5 or 15 minutes).
+         * @return Returns the system load averages.
+         */
+        virtual int get_load_averages(double loads[]) = 0;
```
See above comment.  Should return `std::tuple<double, double, double>`.  Will need to `#include <tuple>`.

peterhuene commented on lib/src/facts/posix/load_average_resolver.cc:
```diff
@@ -0,0 +1,12 @@
+
+    int load_average_resolver::get_load_averages(double loads[])
+    {
+        return(getloadavg(loads, 3));
```
Can do something like:

```c++
array<double, 3> averages;
getloadavg(averages.data(), averages.size());
return make_tuple(averages[0], averages[1], averages[2]);
````

peterhuene replied:
Hmm, although `getloadavg` may not set errno (the documentation is "wonderful").  If so, remove it from the message.

peterhuene commented on lib/schema/facter.yaml:
```diff
@@ -496,6 +496,22 @@ kernelversion:
+    resolution: |
+        Linux: use `getloadavg` function to retrieve the system load averages.
+    elements:
+        load_average_1m:
```
The names of the elements are a little redundant.  Perhaps `1m`, `5m`, and `15m` respectively?

peterhuene commented on lib/src/facts/resolvers/load_average_resolver.cc:
```diff
@@ -0,0 +1,34 @@
+#include <facter/facts/fact.hpp>
+#include <facter/facts/scalar_value.hpp>
+#include <facter/facts/map_value.hpp>
+
```
We generally always do a `using namespace std;` in a .cc file.

peterhuene commented on lib/src/facts/resolvers/load_average_resolver.cc:
```diff
@@ -0,0 +1,34 @@
+        double loads[3];
+
+        /* Get the load averages */
+        if ( get_load_averages(loads) > -1 )
```
With the above changes, this would become:

```c++
auto averages = get_load_averages();
if (averages) {
    auto value = make_value<map_value>();
    value->add("1m",  make_value<double_value>(get<0>(*averages));
    value->add("5m",  make_value<double_value>(get<1>(*averages));
    value->add("15m", make_value<double_value>(get<2>(*averages));
    facts.add(fact::load_averages, move(value));
}

peterhuene commented on lib/tests/facts/schema.cc:
```diff
@@ -138,6 +139,18 @@ struct kernel_resolver : resolvers::kernel_resolver
+ protected:
+    virtual int get_load_averages(double loads[]) override
+    {
+        loads[0] = 0.12;
```
This would simply become:

```c++
return make_tuple(0.12, 3.45, 6.78);
````

peterhuene commented on lib/src/facts/posix/load_average_resolver.cc:
```diff
@@ -0,0 +1,12 @@
+#include <internal/facts/posix/load_average_resolver.hpp>
+#include <stdlib.h>
+
+
```
Same thing here; generally we always do a `using namespace std;` is a cc file.  Generally there shouldn't be any conflicts with libfacter names.

peterhuene commented:
@mattthias Overall the changes look great.  Thanks so much for making this contribution.  I've reviewed your changes and added comments where appropriate.  After the requested changes have been made, I'll get this merged in.  Thanks!

mattthias commented:
Hi @peterhuene , hi @kylog,

i updated the pull-request and incorporated your suggestions. Please review :-)

Best wishes,
Matthias

peterhuene commented on lib/src/facts/resolvers/load_average_resolver.cc:
```diff
@@ -0,0 +1,37 @@
+    {
+    }
+
+void load_average_resolver::resolve(collection& facts)
```
Looks like this line's indentation is off.

peterhuene commented on lib/src/facts/posix/load_average_resolver.cc:
```diff
@@ -0,0 +1,21 @@
+
+namespace facter { namespace facts { namespace posix {
+
+
```
The linter may complain about the double whitespace here, so just keep a look out for any problems with Travis.

peterhuene commented on lib/inc/internal/facts/resolvers/load_average_resolver.hpp:
```diff
@@ -0,0 +1,37 @@
+         */
+        virtual void resolve(collection& facts) override;
+
+    protected:
```
The linter may complain that this isn't indented one space beyond the brackets (it's a little OCD about this stuff), just a heads up.  If Travis passes, then we're good.

peterhuene commented:
@mattthias Other than those minor formatting issues, this looks great.  If you have a python (2, not 3) on your path, you can run "make cpplint" to lint the source to make sure it's clean.  We run the linter as part of Travis, so it could fail CI. Great work! Thanks again for putting this together.

MikaelSmith commented:
Linter failed, as @peterhuene warned about:
`/home/travis/build/puppetlabs/cfacter/lib/src/facts/resolvers/load_average_resolver.cc:23:  Redundant blank line at the start of a code block should be deleted.  [whitespace/blank_line] [2]`

peterhuene commented:
:+1: everything looks good, @mattthias.  One last thing, if you haven't already, could you sign the CLA (same one we use for other PL projects): https://cla.puppetlabs.com/.  You've probably signed it before for other projects, but I need to verify as the bot we use to automatically tag PRs with CLA status isn't working for cfacter :-1: Thanks!

peterhuene commented:
@mattthias Sorry for the spam.  Looks like the CLA has already been signed (figured out how to search its internal database).  Thanks again for the contribution!  I'll get this merged in shortly.

mattthias commented:
Hi @peterhuene, hi @kylog,

thanks for your very constructive help with my pull-request :-). 

best wishes,
Matthias
